### PR TITLE
new: usr: Implementation of email-based OTP

### PR DESCRIPTION
**Pull request by lfortemps**
#### What does it do?

Implementation of #5688 
Using Redis instead of sessions due to volatility.
Introduce new server settings to configure OTP validity, length but also an exclusion list.

#### Questions

- [ ] Does it require a DB change?
- [ ] Are you using it in production?
- [ ] Does it require a change in the API (PyMISP for example)?

#### Release Type:
- [] Major
- [] Minor
- [X] Patch

### CODE REVIEW

File: app/View/Users/email_otp.ctp
Comment (cvandeplas) [inline app/View/Users/email_otp.ctp]: I think this link needs to be generated dynamically. Otherwise the configurable base_url will not work.

File: app/Controller/AppController.php
Comment (cvandeplas) [inline app/Controller/AppController.php]: token should not be on user_id , this would be a vulnerability in the use-case:
- legit user is logged in with OTP-is already validated.
- malicious user logs in with stolen credentials, (s)he does not get the request to the OTP.

File: app/Controller/AppController.php
Comment (cvandeplas) [inline app/Controller/AppController.php]: need to find another way than `$this->_isRest()` to distinguish REST API users. This can be easily circumvented by adding some request header accept response types.  
We might need to review the location of this code. Authentication is already succeeded, and afterwards we do extra authentication elements.

File: app/Model/Server.php
Comment (cvandeplas) [inline app/Model/Server.php]: add a remark it's incompatible with ExternalAuth

File: app/Controller/UsersController.php
Comment (cvandeplas) [inline app/Controller/UsersController.php:35]: should this not be below the self-registration?
As self_registration should work independently of email_otp

File: app/Controller/UsersController.php
Comment (cvandeplas) [inline app/Controller/UsersController.php:1235]: i18n? or is sprintf covered automatically ?

Comment (iglocska) [reply app/Controller/UsersController.php:1235]: Nice catch. Just replace sprintf with __ and all good :)

**Comment by cvandeplas** [comment]
I have no remarks. Just to be sure we'll wait for the green light from Andras before merging

**Review by cvandeplas — APPROVED** [review]

**Comment by SteveClement** [comment]
> 
> 
> I have no remarks. Just to be sure we'll wait for the green light from Andras before merging

@iglocska please ack, or not.

**Pull request merged**
